### PR TITLE
Collection Detail Version Picker

### PR DESCRIPTION
**Pull request by MilanPospisil**
Adds collection version picker to collection detail.

image

### CODE REVIEW

File: frontend/hub/collections/CollectionDetails.tsx [hunk 1]
```diff
@@ -55,20 +56,42 @@ import { HubItemsResponse } from '../useHubView';
...
-  const { data, refresh } = useGet<HubItemsResponse<CollectionVersionSearch>>(
-    hubAPI`/v3/plugin/ansible/search/collection-versions/?name=${params.name || ''}&namespace=${
-      params.namespace || ''
-    }&repository=${params.repository || ''} }`
+  const [searchParams] = useSearchParams();
+  const [collection, setCollection] = useState<CollectionVersionSearch | undefined>(undefined);
+  const itemActions = useCollectionActions(() => void refresh());
+
+  // load all collections versions belong to the repository
+  const { data, refresh, error, isLoading } = useGet<HubItemsResponse<CollectionVersionSearch>>(
```
Comment (nixocio) [inline frontend/hub/collections/CollectionDetails.tsx:71]: 👍

**Review by nixocio — APPROVED** [review]

**Review by vidyanambiar — CHANGES_REQUESTED** [review]
We'll need a space between "updated" and the hours:

Also in the case below (collections i created using galaxykit), all the version numbers in the picker are the same and the 2nd, 3rd, and and 4th entries look identical. Is that normal? 



Screenshot 2023-09-01 at 9 10 18 AM

**Comment by himdel** [comment]
Don't forget about the version picker *modal* as well, when there is more than 10 version, the picker should be showing the first 10 and a button to open the modal with pagination.

**Comment by MilanPospisil** [comment]
> 

Ok, than I will finish my changes to Framework first, this will simplify those modals.

File: frontend/hub/collections/CollectionDetails.tsx [hunk 2]
```diff
@@ -55,20 +56,42 @@ import { HubItemsResponse } from '../useHubView';
...
+    hubAPI`/v3/plugin/ansible/search/collection-versions/?name=${
+      searchParams.get('name') || ''
+    }&namespace=${searchParams.get('namespace') || ''}&repository_name=${
+      searchParams.get('repository') || ''
```
Comment (himdel) [inline frontend/hub/collections/CollectionDetails.tsx:75]:
```suggestion
       collection?.repository.name || ''
```

maybe?

Comment (MilanPospisil) [reply frontend/hub/collections/CollectionDetails.tsx:75]: Collection is selected collection. That would not work...

File: frontend/hub/collections/CollectionDetails.tsx [hunk 3]
```diff
@@ -84,7 +107,59 @@ export function CollectionDetails() {
             selectedItem={collection}
           />
         }
+        description={t('Repository: ') + collection?.repository.name}
```
[the remark this answers is not in the document]
Comment (MilanPospisil) [reply frontend/hub/collections/CollectionDetails.tsx:110]: Do you think repository can be null? Hmm... I should have generate all inner object of API interfaces with nullables, so typescript will always catch those. We cant rely that the API returns what it should return :-)

**Comment by MilanPospisil** [comment]
> We'll need a space between "updated" and the hours:
> 
> Also in the case below (collections i created using galaxykit), all the version numbers in the picker are the same and the 2nd, 3rd, and and 4th entries look identical. Is that normal?
> 
> Screenshot 2023-09-01 at 9 10 18 AM

I cant replicate the problem. Do you have repository in URL? The query does have repository filter and I have multiple versions in multiple repos and I dont see them in one dropdown.

**Comment by MilanPospisil** [comment]
I apologize for mayor change. I forgot about version in URL - sometimes we need to redirect to certain version, rather than select it. Now the selector dropdown changes URL and version is loaded using URL parameters. If URL param version is missing, it will use is_highest instead of version. It can load version directly without any selection if version is in URL.

File: frontend/hub/collections/CollectionDetails.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (himdel) [reply frontend/hub/collections/CollectionDetails.tsx:162]: Wasn't `null` the behavior for orphaned collection versions? But maybe those get filtered out now, I'm not sure.

File: frontend/hub/collections/CollectionDetails.tsx (hunk 2 above)
Comment (himdel) [reply frontend/hub/collections/CollectionDetails.tsx:83]: Why not? The repo of the current collection sounds like the right thing to filter by when searching for other versions of the same collection in the same repo..

(but maybe I'm misunderstanding `searchParams`? I thought it was a query string param, but maybe it's a route param?)

File: frontend/hub/collections/CollectionDetails.tsx [hunk 4]
```diff
@@ -55,20 +56,92 @@ import { HubItemsResponse } from '../useHubView';
...
+  } else {
+    versionFilter = '&version=' + version;
+  }
+
+  const { data, refresh, error, isLoading } = useGet<HubItemsResponse<CollectionVersionSearch>>(
+    hubAPI`/v3/plugin/ansible/search/collection-versions/?name=${
+      searchParams.get('name') || ''
+    }&namespace=${searchParams.get('namespace') || ''}&repository_name=${
+      searchParams.get('repository') || ''
+    }&order_by=-pulp_created` +
```
Comment (himdel) [inline frontend/hub/collections/CollectionDetails.tsx:118]: (and here)

**Review by vidyanambiar — APPROVED** [review]
lgtm!

File: frontend/hub/collections/CollectionDetails.tsx (hunk 2 above)
Comment (MilanPospisil) [reply frontend/hub/collections/CollectionDetails.tsx:83]: Never mind, I added version to search params, so now collection is loaded based on search params, not based on selection (selection changes search params version and collection is loaded then).

File: frontend/hub/collections/CollectionDetails.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (MilanPospisil) [reply frontend/hub/collections/CollectionDetails.tsx:162]: You are maybe right. Im just saying that we should first change the definition in repository if this is true, so repository is nullable, that will enforce any check automaticaly.

[the remark this answers is not in the document]
Comment (himdel) [reply frontend/hub/collections/CollectionDetails.tsx:162]: Yeah, I was mostly just pointing out that `foo?.bar.baz` doesn't make sense and should be either `foo.bar.baz` or `foo?.bar?.baz` :)

[the remark this answers is not in the document]
Comment (MilanPospisil) [reply frontend/hub/collections/CollectionDetails.tsx:162]: Yeah, thats why we should do API results definitions with nullable ?, so typescript will enforce it.

[the remark this answers is not in the document]
Comment (MilanPospisil) [reply frontend/hub/collections/CollectionDetails.tsx:162]: Lets merge this and I will do it in seperate PR.

File: frontend/hub/collections/CollectionDetails.tsx (hunk 4 above)
Comment (MilanPospisil) [reply frontend/hub/collections/CollectionDetails.tsx:118]: I hope this will always return 1 item anyway :)

**Pull request merged**